src/AutoRest.CSharp/Common/Output/Models/Types/HelperTypeProviders/ChangeTrackingListProvider.cs

```diff
@@ -0,0 +1,282 @@
+            _tArray = new GenericParameterTypeProvider("T[]", DefaultNamespace, null);
+            _tParam = new Parameter("item", null, _t.Type, null, ValidationType.None, null);
+            DeclarationModifiers = ClassSignatureModifiers.Internal;
+        }
+
+        protected override string DefaultName => "ChangeTrackingList";
+
+        protected override IEnumerable<Method> BuildConstructors()
+        {
+            yield return new Method(new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, Array.Empty<Parameter>()), Snippets.EmptyStatement);
+            var iListParam = new Parameter("innerList", null, typeof(IList<>), null, ValidationType.None, null);
+            var iListSignature = new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, new Parameter[] { iListParam });
+            var iListVariable = new ParameterReference(iListParam);
+            var iListbody = new MethodBodyStatement[]
+            {
+                new IfStatement(Snippets.NotEqual(iListVariable, Snippets.Null))
+                {
+                    new AssignValueStatement(_innerList, iListVariable)
+                }
+            };
+
+            yield return new Method(iListSignature, iListbody);
+            var iReadOnlyListParam = new Parameter("innerList", null, typeof(IReadOnlyList<>), null, ValidationType.None, null);
+            var iReadOnlyListSignature = new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, new Parameter[] { iReadOnlyListParam });
+            var iReadOnlyListVariable = new ParameterReference(iReadOnlyListParam);
+            var iReadOnlyListbody = new MethodBodyStatement[]
+            {
+                new IfStatement(Snippets.NotEqual(iReadOnlyListVariable, Snippets.Null))
+                {
+                    new AssignValueStatement(_innerList, Snippets.Linq.ToList(iReadOnlyListVariable))
+                }
+            };
+
+            yield return new Method(iReadOnlyListSignature, iReadOnlyListbody);
+        }
+
+        protected override IEnumerable<CSharpType> BuildTypeArguments()
+        {
+            yield return _t.Type;
+        }
+
+        protected override IEnumerable<CSharpType> BuildImplements()
+        {
+            yield return new CSharpType(typeof(IList<>));
+            yield return new CSharpType(typeof(IReadOnlyList<>));
+        }
+
+        protected override IEnumerable<FieldDeclaration> BuildFields()
+        {
+            yield return _innerlListField;
+        }
+
+        protected override IEnumerable<PropertyDeclaration> BuildProperties()
+        {
+            yield return new PropertyDeclaration(null, MethodSignatureModifiers.Public, typeof(bool), "IsUndefined", new ExpressionPropertyBody(Snippets.Equal(_innerList, Snippets.Null)));
+            yield return BuildCount();
+            yield return BuildIsReadOnly();
+            yield return BuildIndexer();
```

Comment:
I think we should make these properties as fields or lazy loaded properties of this class because we are using them in other places, for example, the property `IsUndefined` is used in a lot of methods and other properties.
So that we could just reference the `PropertyDeclaration` instance instead of hard-coding a lot of strings (like `IsUndefined`) everywhere, we only need to put the hard-code string in one place.

In the meantime, we have an implicit operator from `FieldDeclartion` to `ValueExpression`, therefore I think we could have one for `PropertyDeclaration` as well.